:zap: avoid expensive & unnecessary groupby in px

jvdd opened the pull request:
This PR enables significant speed-ups for `plotly.express` when no [groupby operation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html) is required. This PR stems from the observations I made in #1743; current behavior always performs at least one groupby, even when this is not necessary (e.g., `px.scatter` of a 1D array).

Performing such a groupby is a rather expensive operation + the `.get_group` and the `get_ordenings` operations are also rather time-consuming. This PR checks via the `all_same` variable whether grouping is required. If no grouping is necessary, no groupby will be performed, resulting in significant performance gains.

### Illustration of the speed-up

**plotly 5.8.0**
![image](https://user-images.githubusercontent.com/18898740/172217572-dcfa68b5-7cdb-42a3-b8f1-1c62e1290fc8.png)

**this PR**
![image](https://user-images.githubusercontent.com/18898740/172218251-9431683d-b4c8-4dda-965b-482ea02593bb.png)

Can you validate that this speed-up is correct? 
(FYI: all pytest tests succeed locally)

## Code PR checklist

- [x] I have read through the [contributing notes](https://github.com/plotly/plotly.py/blob/master/contributing.md) and understand the structure of the package. In particular, if my PR modifies code of `plotly.graph_objects`, my modifications concern the `codegen` files and not generated files.
- [ ] I have added tests (if submitting a new feature or correcting a bug) or modified existing tests.
- [ ] For a new feature, I have added documentation examples in an existing or new tutorial notebook (please see the doc checklist as well).
- [ ] I have added a CHANGELOG entry if fixing/changing/adding anything substantial.
- [ ] For a new feature or a change in behaviour, I have updated the relevant docstrings in the code to describe the feature or behaviour (please see the doc checklist as well).

nicolaskruchten commented:
Thanks for this PR, it looks like a promising optimization!

I certainly like the idea of skipping the grouping when it's all `one_group`... When plotting, say, a 5000-row dataframe with 20 colors and 5 symbols or something, what is the performance hit here of the nunique call?

jvdd commented:
Hi @nicolaskruchten 

In the latest commit I did some minor refactoring (renaming variables) + improved the efficiency of the check I added (replaced `.nunique` with the suggestion from here https://stackoverflow.com/a/54405767

I checked the performance for the following plot; (is this what you intended?)

```py
import pandas as pd
import numpy as np
import plotly.express as px

df = pd.DataFrame(np.random.randn(5_000, 1), columns=["data"])
df["c"] = np.tile(np.arange(20), 5_000 // 20)  # 20 colors
df["s"] = np.tile(np.arange(5), 5_000 // 5)  # 5 symbols

fig = px.scatter(df, y="data", color="c", symbol="s")
```

The checks that I added for grouping (i.e., creating the `all_same_group` variable) take 0.1-0.2% of the total time of `make_figure`.
```py
%lprun -f px._core.make_figure px.scatter(df, y="data", color="c", symbol="s")
```

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -1955,10 +1955,37 @@ def make_figure(args, constructor, trace_patch=None, layout_patch=None):
+        # no groups to group by, or when the group has only one (i.e., the same) value
+        grouper = [g for g in grouper_ if g is not one_group]
+        assert len(grouper) <= 1
+        # -> create orders, sorted_group_names equivalent to those of get_ordings
```
I love the optimization where we don't group if the groupers are all `one_group` but I'm worried about the duplication of logic here... If we just do the optimization when no grouping is requested at all (i.e. leave out the case when we do have e.g. `symbol` but there's only one value) then this PR can be less invasive/have less duplication. I've done an implementation of that over here https://github.com/plotly/plotly.py/pull/3765 if you want to take a look at it.

I think that most of the value of this PR comes from the "all `one_group`" case, and that leaving out the case where there happens to only be one group would be OK for now as this is likely not all that common, but I'm not dead set on it.

jvdd replied:
Do you mean like your implementation in #3765?

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -1955,10 +1955,37 @@ def make_figure(args, constructor, trace_patch=None, layout_patch=None):
+        # no groups to group by, or when the group has only one (i.e., the same) value
+        grouper = [g for g in grouper_ if g is not one_group]
+        assert len(grouper) <= 1
+        # -> create orders, sorted_group_names equivalent to those of get_ordings
```
I love the optimization where we don't group if the groupers are all `one_group` but I'm worried about the duplication of logic here... If we just do the optimization when no grouping is requested at all (i.e. leave out the case when we do have e.g. `symbol` but there's only one value) then this PR can be less invasive/have less duplication. I've done an implementation of that over here https://github.com/plotly/plotly.py/pull/3765 if you want to take a look at it.

I think that most of the value of this PR comes from the "all `one_group`" case, and that leaving out the case where there happens to only be one group would be OK for now as this is likely not all that common, but I'm not dead set on it.

nicolaskruchten commented:
Superseded by #3765
